### ISSUE TITLE
lazzy_attribute with Faker('date_between_dates', ...).generate({...}) return date of today

### ISSUE THREAD

**Issue by FractalWire** [FactoryBoy/factory_boy#815]
#### Description

When using Faker('date_between_dates', ...).generate(...) on a `lazzy_attribute` definition, the generated date is the date of today instead of a date between `start_date` and `end_date`

##### Model / Factory code
```python

@dataclass(frozen=True)
class Search:
    id: int
    latitude: InitVar[float]
    longitude: InitVar[float]
    start_date: datetime = None
    end_date: datetime = None

from random import uniform
from datetime import date

import factory

xmin, xmax, ymin, ymax = (-180,180,-90,90,)
coordinates_precision = 5
price_precision = 2
daterange = (
    date.fromisoformat('2020-01-01'),
    date.fromisoformat('2021-01-01'),
)

fake = factory.Faker

class SearchFactory(factory.Factory):
    class Meta:
        model = Search

    id = factory.Sequence(lambda n: n)
    latitude = fake('pyfloat', min_value=ymin,
                    max_value=ymax, right_digits=coordinates_precision)
    longitude = fake('pyfloat', min_value=xmin,
                     max_value=xmax, right_digits=coordinates_precision)
    start_date = fake(
        'date_between_dates',
        date_start=daterange[0],
        date_end=daterange[1]
    )

    # /!\ Faker is not working properly here -> return date of today instead /!\
    end_date = factory.LazyAttribute(lambda search: fake(
        'date_between_dates',
        date_start=daterange[0],
        date_end=daterange[1]
    ).generate({'locale': None})
    )

```

##### The issue

```python
>>> from tests.models_factory import SearchFactory
>>> SearchFactory()
Search(id=0, point=Point(x=-35.89168, y=-52.98786), start_date=datetime.date(2020, 4, 15), end_date=datetime.date(2020, 11, 5))
>>> SearchFactory()
Search(id=1, point=Point(x=-147.57452, y=56.5883), start_date=datetime.date(2020, 1, 17), end_date=datetime.date(2020, 11, 5))
>>> SearchFactory()
Search(id=2, point=Point(x=-51.26297, y=-34.1516), start_date=datetime.date(2020, 2, 24), end_date=datetime.date(2020, 11, 5))
```

`end_date=datetime.date(2020, 11, 5)` for every generated Search, which is today...

#### Notes

Using `fuzzy.FuzzyDate` works well however which allow to do something like that:

```python
    @factory.lazy_attribute
    def end_date(self):
        return factory.fuzzy.FuzzyDate(
            self.start_date,
            daterange[1]
        ).fuzz()
```

**Comment by rbarrois** [comment]
Hi, and thanks for the report!

I'm a bit surprised though: in your code, `end_date` seems to use the exact same computation as `start_date`, why wouldn't you use a `factory.Faker()` call directly?

`factory.Faker().generate()` is a private API, and not intended for direct use...

**Comment by FractalWire** [comment]
Indeed, I should have outline my real use case for that which is more in line with what I provided with the fuzzy alternative:

```python
    end_date = factory.LazyAttribute(lambda search: fake(
        'date_between_dates',
        date_start=search.start_date,
        date_end=daterange[1]
    ).generate({'locale': None})
    )
```

This ends with the same result as a simple `date_start=daterange[0],`, which I guess means the output will be the same no matter the arguments that are passed to Faker()

I must use `generate()` (or `fuzz()` for the fuzzy alternative) here otherwise I'd have an output like so:

```python
>>> SearchFactory()
Search(id=0, point=Point(x=-86.89367, y=-31.14752), start_date=datetime.date(2020, 9, 28), end_date=<factory.faker.Faker object at 0x7f6e9fafcd60>)
```

Maybe you have an alternative way not using `generate()` or `fuzz()` ?

**Comment by rbarrois** [comment]
Since #772 was merged (in 3.1.0), this is now possible:
```python
end_date = factory.Faker(
    'date_between_dates',
    date_start=factory.SelfAttribute('..start_date'),
    date_end=daterange[1],
)
```

See the example in the [`Faker docs`](https://factoryboy.readthedocs.io/en/stable/reference.html#faker) ;)

**Comment by rbarrois** [comment]
I feel this has solved your issue, so I'm closing it; feel free to reopen it if needed ;)
